src/gmp_backend.rs

```diff
@@ -256,9 +269,59 @@ impl Bn {
         }
     }
 
+    /// Generate a safe prime with `size - 1` or `size` bits using a user-provided rng
+    pub fn fast_safe_prime_with_rng(rng: &mut (impl CryptoRng + RngCore), bitsize: usize) -> Self {
+        let mut p = _random_nbit(rng, bitsize - 1);
+
+        // Set the MSB bit so that we're sampling from [2^(size - 2), 2^(size - 1))
+        p.setbit(bitsize - 2);
+
+        loop {
+            // TODO: Safe primes are 2 (mod 3). We don't need to generate primes that are 1 (mod 3)
+            // orig_p
+            p = p.nextprime();
+
+            // 2 orig_p + 1
+            p <<= 1;
+            p += 1;
+
+            // Keep the number of Miller-Rabin rounds high since
+            // a non-negligible false positive will affect key recovery
+            if let ProbabPrimeResult::Prime | ProbabPrimeResult::ProbablyPrime = p.probab_prime(25)
+            {
+                return Self(p);
+            }
+
+            // Skip the following if (orig_p - 1)/2 is even
+            if !p.tstbit(2) {
+                p >>= 1;
+                continue;
+            }
+
+            // (orig_p - 1)/2
+            p >>= 2;
+            let is_prime_res = p.probab_prime(25);
+
+            // orig_p
+            p <<= 1;
+            p += 1;
+
+            if let ProbabPrimeResult::Prime | ProbabPrimeResult::ProbablyPrime = is_prime_res {
+                return Self(p);
```

Comment:
In this code path we return a safe prime whose bit length is 1 less than the requested `bitsize`.  Is that a ok?

Reply:
Yep, few bits of difference doesn't really affect Paillier (the concern is that `p | q-1` or `q | p-1` shouldn't occur) and the optimization is worth it. I have mentioned this behaviour in the doc above and will also more info in the paillier library.